File: algorithms/javaalgo/src/test/java/programmers/level2/tangerinechoice/TangerineChoiceTest.java
package programmers.level2.tangerinechoice;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

public class TangerineChoiceTest {

    @Test
    void test() {
        int k1 = 6;
        int[] tangerine1 = {1, 3, 2, 5, 4, 5, 2, 3};

        int k2 = 4;
        int[] tangerine2 = {1, 3, 2, 5, 4, 5, 2, 3};

        int k3 = 2;
        int[] tangerine3 = {1, 1, 1, 1, 2, 2, 2, 3};

        assertAll(
            () -> assertThat(solution(k1, tangerine1)).isEqualTo(3),
            () -> assertThat(solution(k2, tangerine2)).isEqualTo(2),
            () -> assertThat(solution(k3, tangerine3)).isEqualTo(1)
        );
    }

    public int solution(int k, int[] tangerine) {
        int answer = 0;
        Map<Integer, Integer> maps = new HashMap<>();
        for (int i : tangerine) {
            maps.put(i, maps.getOrDefault(i, 0) + 1);
        }
        List<Integer> counts = new ArrayList<>(maps.values());
        counts.sort(Collections.reverseOrder());

        for (int count : counts) {
            if (k > 0) {
                k -= count;
                answer++;
            } else {
                return answer;
            }
        }

        return answer;
    }
}

File: algorithms/javascriptalgo/programmers/level2/tangerineChoice.test.js
test("test", () => {
    let k1 = 6;
    let tangerine1 = [1, 3, 2, 5, 4, 5, 2, 3];

    let k2 = 4;
    let tangerine2 = [1, 3, 2, 5, 4, 5, 2, 3];

    let k3 = 2;
    let tangerine3 = [1, 1, 1, 1, 2, 2, 2, 3];

    expect(solution(k1, tangerine1)).toBe(3);
    expect(solution(k2, tangerine2)).toBe(2);
    expect(solution(k3, tangerine3)).toBe(1);
});

function solution(k, tangerine) {
    let answer = 0;
    const maps = {};
    // 함수형으로 Map으로 ++ 하기
    tangerine.forEach((t) => maps[t] = (maps[t] || 0) + 1);
    const counts = Object.values(maps).sort((a, b) => b - a);
    for (const count of counts) {
        if (k > 0) {
            k -= count;
            answer++;
        } else {
            return answer;
        }
    }

    return answer;
}
